proposals/0317-async-let.md

```diff
@@ -635,56 +640,233 @@ As discussed in the [structured concurrency proposal](0304-structured-concurrenc
 
 ```swift
-func hello(guest name: String) async -> String {
-  async let registered = register(name: name, delayInSeconds: 3) // error: `registered` was not awaited on
-  // ... 
-  return "Hello \(name)!"
+func runLoop() async {
+  for e in list {
+    async let a = f(e)
+    guard <condition> else {
+      await break   // awaits the child task that produces the value a
+    }
+    ... await a ...
+  }
+  foo()
 }
 ```
 
-It does not scale well to more realistic functions that include multiple branches and more complicated control flow:
+One would similarly need an `await continue`. For the second example, this means marking the call to `mayFail()` with an `await`, because the potentially-throwing call creates a control-flow edge out of the scope:
 
 ```swift
-func maybeHello(guest name: String, greetLoudly: Bool) async -> String { 
-  async let registered = register(name: name, delayInSeconds: 3)
+func runException() async {
+  do {
+    async let a = f()
+    try await mayFail() // awaits the child task that produces a; mayFail() itself may not even be "async"
+    ... await a ...
+  } catch {
+    ...
+  }
+}
+```
+
+Similarly, one would need `await throw` for cases where a directly-thrown expression would imply a suspension point to wait for an `async let` child task to complete:
+
+```swift
+func runThrow() async {
+  do {
+    async let a = f()
+    if <condition> {
+      await throw SomeError() // awaits the child task that produces a
+    }
+    ... await a ...
+  } catch {
+    ...
+  }
+}
+```
+
+However, not all control-flow edges involving implicit `async let` suspension points have a specific keyword to which we can attach `await`, because some come from fall-through to subsequent code. For such cases, one could have a standalone `await` statement marking that fall through:
+
+```swift
+func runIfFallthrough() async {
+  if <condition> {
+    async let a = f()
+    ... code ...
+    // falling out of this block must await the child task that produces a, so require a freestanding "await"
+    await
+  }
+  ... more code ...
+}
+```
+
+The same would be required in, e.g., the cases of a `switch` statement that introduce an `async let`:
+
+```swift
+func runSwitchCase() async {
+  switch <expression> {
+  case .a:
+    async let a = f()
+    // falling out of this block must await the child task that produces a, so require a freestanding "await"
+    await
+
+  default:
+    ... code ...
+  }
+  ... code ...
+}
+```
+
+The above is a significant expansion of the grammar: introducing the `await` keyword in front of `break`, `continue`, `throw`, and `fallthrough`; requiring `await` on certain throwing expressions; and adding the freestanding `await`statement. It would also need to be coupled with rules that only require the new `await` when it is semantically meaningful. For example, the additional `await` shouldn't be required if all of the `async let` child tasks have already been explicitly awaited in some other manner, e.g.,
```

Comment:
```suggestion
The above is a significant expansion of the grammar: introducing the `await` keyword in front of `break`, `continue`, `throw`, and `fallthrough`; requiring `await` on certain throwing expressions; and adding the freestanding `await` statement. It would also need to be coupled with rules that only require the new `await` when it is semantically meaningful. For example, the additional `await` shouldn't be required if all of the `async let` child tasks have already been explicitly awaited in some other manner, e.g.,
```